lib/persist-state.c

```diff
@@ -889,26 +890,43 @@ void
 persist_state_alloc_string(PersistState *self, const gchar *persist_name, const gchar *value, gssize len)
 {
   PersistEntryHandle handle;
-  SerializeArchive *sa;
+  SerializeArchive *serializer;
   GString *buf;
+  gsize buflen;
   gboolean success;
   gpointer block;
 
+  PersistEntryHandle lookup_handle;
+  gsize lookup_size;
+  guint8 lookup_result_version;
+  gpointer lookup_block;
+
   if (len < 0)
     len = strlen(value);
 
-  buf = g_string_sized_new(len + 5);
-  sa = serialize_string_archive_new(buf);
-
-  success = serialize_write_cstring(sa, value, len);
+  buflen = len + 5;
+  buf = g_string_sized_new(buflen);
+  serializer = serialize_string_archive_new(buf);
+  success = serialize_write_cstring(serializer, value, len);
   g_assert(success == TRUE);
+  serialize_archive_free(serializer);
 
-  serialize_archive_free(sa);
+  lookup_handle = persist_state_lookup_entry(self, persist_name, &lookup_size, &lookup_result_version);
+
+  if (lookup_handle && lookup_size >= buflen)
+    {
+      lookup_block = persist_state_map_entry(self, lookup_handle);
+      memcpy(lookup_block, buf->str, buf->len);
+      persist_state_unmap_entry(self, lookup_handle);
+    }
+  else
+    {
+      handle = persist_state_alloc_entry(self, persist_name, buf->len);
+      block = persist_state_map_entry(self, handle);
+      memcpy(block, buf->str, buf->len);
+      persist_state_unmap_entry(self, handle);
+    }
```

Comment:
Optional:

You can reduce a bit of duplication by making only the allocation conditional, for example:

```
PersistEntryHandle handle = persist_state_lookup_entry(...);
if (!handle || lookup_size < buflen)
  handle = persist_state_alloc_entry(self, persist_name, buf->len);

gpointer block = persist_state_map_entry(self, handle);
memcpy(...);
persist_state_unmap_entry(...);
```

Reply:
I think this would worth fixing though.